### PR TITLE
[Evaluation] Modularization & enabling custom evaluations

### PR DESCRIPTION
**Pull request by kaisopos**
# Description

[Evaluation] Modularization & enabling custom evaluations.

The most important changes introduced by this XL PR are: 
- Made our evaluation object-oriented by defining an Evaluator class
- Defined clean interfaces among eval modules (eg. `EvaluationResult` shared across all backends)
- Added a new backend: "CUSTOM"
- Enabled users to register their own evaluation functions using the new custom backend
- Refactored lm harness, alpaca eval (as backend convenience functions)
- Re-wrote all evaluation unit tests (which were problematic)
- Added new registry unit tests, and unit tests for all new functionality
- Addressed a few corner cases that we had missed before (eg. Saving to an existing dictionary is now NOT allowed, but a new indexed dictionary is automatically generated)
- Renamed `evaluation_platform`s (LM Harness, Alpaca Eval) -> `evaluation_backend` (**_breaking change_**)
- Dropped `CustomEvaluationParams` which was not used

## Related issues

Towards OPE-1097

## Before submitting

- [ ] This PR only changes documentation. (You can ignore the following checks in that case)
- [x] Did you read the [contributor guideline](https://github.com/oumi-ai/oumi/blob/main/CONTRIBUTING.md) Pull Request guidelines?
- [x] Did you link the issue(s) related to this PR in the section above?
- [x] Did you add / update tests where needed?

## Reviewers

At least one review from a member of `oumi-ai/oumi-staff` is required.

### CODE REVIEW

File: tests/unit/core/evaluation/test_backend_lm_harness.py [hunk 1]
```diff
@@ -1,46 +1,46 @@
+import copy
 from unittest.mock import MagicMock, patch
 
 import pytest
 import torch
 from lm_eval.api.task import ConfigurableTask
```
Comment (taenin) [inline tests/unit/core/evaluation/test_backend_lm_harness.py:7]: In the test run, I see:
```
22.99s call     tests/unit/core/evaluation/test_backend_lm_harness.py::test_evaluate
```
https://github.com/oumi-ai/oumi/actions/runs/13582033201/job/37969655852
That indicates to me that you're not mocking out all LM Harness dependencies in this test.

I would expect this test to take <1s

File: src/oumi/core/registry/registry.py [hunk 2]
```diff
@@ -186,6 +188,10 @@ def get_judge_config(self, name: str) -> Optional[Callable]:
         """Gets a record that corresponds to a registered judge config."""
         return self.get(name, RegistryType.JUDGE_CONFIG)
 
+    def get_evaluate_function(self, name: str) -> Optional[Callable]:
```
Comment (taenin) [inline src/oumi/core/registry/registry.py:191]: Per the description this should be called `get_evaluation_function`

File: src/oumi/core/registry/registry.py [hunk 3]
```diff
@@ -323,3 +329,50 @@ def decorator_register(obj):
         return obj
 
     return decorator_register
+
+
+def register_evaluate_function(registry_name: str) -> Callable:
```
Comment (taenin) [inline src/oumi/core/registry/registry.py:334]: `register_evaluation_function`

File: src/oumi/core/registry/registry.py [hunk 4]
```diff
@@ -323,3 +329,50 @@ def decorator_register(obj):
...
+    """Returns function to register an evaluation function in the Oumi global registry.
+
+    Args:
+        registry_name: The name that the evaluation function should be registered with.
+
+    Returns:
+        Decorator function to register the target evaluation function.
+    """
+
+    def check_evaluate_function_signature(evaluate_fn):
```
Comment (taenin) [inline src/oumi/core/registry/registry.py:344]: `check_evaluation_function_signature`

File: src/oumi/core/registry/registry.py [hunk 5]
```diff
@@ -323,3 +329,50 @@ def decorator_register(obj):
...
+                f"`{RegistryType.EVALUATE_FUNCTION}` must be callable."
+            )
+
+        signature = inspect.signature(evaluate_fn)
+        if (
+            "task_params" not in signature.parameters
+            or "CustomTaskParams" not in str(signature.parameters["task_params"])
+            or "config" not in signature.parameters
+            or "EvaluationConfig" not in str(signature.parameters["config"])
+            or "EvaluationResult" not in str(signature.return_annotation)
```
Comment (taenin) [inline src/oumi/core/registry/registry.py:358]: A few high level comments: Do you think it's a heavy lift to assert typing. Do you, as a researcher, often put typing in your functions? 

If you do check typing, though, I think checking `EvaluationTaskParams` is better than `CustomTaskParams`, as `CustomTaskParams` doesn't give them anything useful during execution.

File: src/oumi/core/registry/registry.py [hunk 6]
```diff
@@ -32,6 +33,7 @@ class RegistryType(Enum):
     MODEL_CONFIG = auto()
     MODEL = auto()
     JUDGE_CONFIG = auto()
+    EVALUATE_FUNCTION = auto()
```
Comment (taenin) [inline src/oumi/core/registry/registry.py:36]: Made this comment in many other places, but let's change `Evaluate` -> `Evaluation`

File: src/oumi/core/registry/registry.py (hunk 2 above)
Comment (kaisopos) [reply src/oumi/core/registry/registry.py]: Fixed

File: src/oumi/core/registry/registry.py (hunk 3 above)
Comment (kaisopos) [reply src/oumi/core/registry/registry.py]: Fixed

File: src/oumi/core/registry/registry.py (hunk 6 above)
Comment (kaisopos) [reply src/oumi/core/registry/registry.py]: Fixed

**Comment by wizeng23** [comment]
This is a pretty large change. Could you add a breakdown of the PR in your PR description?

File: src/oumi/core/registry/registry.py (hunk 5 above)
Comment (kaisopos) [reply src/oumi/core/registry/registry.py]: Thanks for pointing this out!
Absolutely, no researcher will ever declare types (I really did not think of that).
Thank you - fixed!

File: src/oumi/core/configs/params/evaluation_params.py [hunk 7]
```diff
@@ -14,26 +14,26 @@
...
 
 
 @dataclass
 class EvaluationTaskParams(BaseParams):
     """Configuration parameters for model evaluation tasks.
 
-    Supported platforms:
+    Supported backends:
 
     - LM Harness: Framework for evaluating language models on standard benchmarks.
```
Comment (wizeng23) [inline src/oumi/core/configs/params/evaluation_params.py:38]: Could you also add mention of the custom backend here?

File: src/oumi/core/configs/params/evaluation_params.py
Comment (wizeng23) [inline src/oumi/core/configs/params/evaluation_params.py:1]: If you haven't, could you do a quick ctrl+f throughout the repo for any other mentions of "platform", "evaluation platform", "evaluation_platform", etc.?

File: src/oumi/core/evaluation/backends/alpaca_eval.py [hunk 8]
```diff
@@ -167,30 +155,19 @@ def evaluate(
...
+            "model_outputs": responses_json,
+        }
+
+        return EvaluationResult(
+            task_name=task_params.task_name,
+            task_result={"results": metric_dict},
+            backend_config=backend_task_config,
+        )
+
+    return EvaluationResult()
```
Comment (wizeng23) [inline src/oumi/core/evaluation/backends/alpaca_eval.py:173]: nit: A small tip to improve readability: To reduce the overall amount of indentation in the code and make it more readable, you can invert the if case to have the smaller code block first. For example:

```
if not is_world_process_zero():
    return EvaluationResult()
metric_dict = {}
...
return EvaluationResult(...)
```

Comment (wizeng23) [reply src/oumi/core/evaluation/backends/alpaca_eval.py:173]: While it's minor, code readers need to track additional things in their head when they enter if blocks. By making the if block as short as possible, it's a bit simpler

File: src/oumi/core/evaluation/backends/lm_harness.py [hunk 9]
```diff
@@ -371,7 +353,6 @@ def evaluate(
         apply_chat_template=is_multimodal,
         **task_params.eval_kwargs,  # type: ignore
     )
-    elapsed_time_sec = time.time() - start_time
 
     # Metrics are only available on the main process, and `None` on others.
     if is_world_process_zero():
```
Comment (wizeng23) [inline src/oumi/core/evaluation/backends/lm_harness.py:358]: ditto, but an even better example here, as there's 50 indented lines here in the if block

File: src/oumi/core/evaluation/evaluation_result.py [hunk 10]
```diff
@@ -0,0 +1,36 @@
...
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from typing import Any, Optional
+
+
+class EvaluationResult:
```
Comment (wizeng23) [inline src/oumi/core/evaluation/evaluation_result.py:18]: Can this be a dataclass?

**Comment by wizeng23** [comment]
IMO this PR should be broken up. 1k lines of non-trivial code logic changes is very hard to digest for reviewers not familiar with the subject. Just so I'm not hypocritical, I admit I'm also guilty of this :') However, I think it's very important for big logic refactors like this to be carefully reviewed, and this is currently too big for me to wrap my head around in one review (especially without a PR description of what's going on).

AFAIK there's a lot of distinct code logic changes that don't depend on each other, so for example you could do:
1. Rename platform -> backend throughout
2. Refactor evaluate() functions for LM Harness and Alpaca to take the EvaluationConfig as input
3. Add new classes like Evaluator, EvaluationResult, etc.
4. Swap over from the old logic to the new logic
5. Add the custom evaluation backend

This is just a guess, so feel free to do whatever you think is best.

Breaking up big PRs is pretty hard and annoying though; feel free to DM me if you want advice for it as I've had to do it before.

File: tests/unit/core/evaluation/test_backend_alpaca_eval.py [hunk 11]
```diff
@@ -15,7 +15,6 @@
     GenerationParams,
     ModelParams,
 )
-from oumi.evaluation.save_utils import OUTPUT_FILENAME_PLATFORM_RESULTS
```
[the remark this answers is not in the document]
Comment (kaisopos) [reply tests/unit/core/evaluation/test_backend_alpaca_eval.py:18]: Fixed

File: src/oumi/core/configs/params/evaluation_params.py [hunk 12]
```diff
@@ -81,44 +81,48 @@ class EvaluationTaskParams(BaseParams):
...
+        elif self.evaluation_backend == EvaluationBackend.CUSTOM.value:
+            return EvaluationBackend.CUSTOM
         else:
-            raise ValueError(f"Unknown evaluation platform: {self.evaluation_platform}")
+            raise ValueError(f"Unknown evaluation backend: {self.evaluation_backend}")
 
-    def get_evaluation_platform_task_params(self):
-        """Returns the evaluation platform-specific task parameters."""
-        if self.get_evaluation_platform() == EvaluationPlatform.LM_HARNESS:
+    def get_evaluation_backend_task_params(self):
```
[the remark this answers is not in the document]
Comment (kaisopos) [reply src/oumi/core/configs/params/evaluation_params.py:103]: These classes are defined later in the file (since they inherit from the current class [EvaluationTaskParams]).
This method is kind of a factory pattern, so I could extract it and have it as a stand-alone function in the file, if you think it's cleaner this way.

File: tests/unit/core/evaluation/test_backend_lm_harness.py (hunk 1 above)
Comment (kaisopos) [reply tests/unit/core/evaluation/test_backend_lm_harness.py:7]: Fixed here: https://github.com/oumi-ai/oumi/actions/runs/13598997772/job/38021806619?pr=1484

Slowest duration in "slowest 50 durations" is 1.55s.
No evaluation test is >0.5sec

File: src/oumi/core/evaluation/evaluator.py [hunk 13]
```diff
@@ -0,0 +1,190 @@
...
+        """Retrieve the evaluation function of the custom task."""
+        if not task_params.task_name:
+            raise ValueError(
+                "Missing `task_name` for custom Oumi evaluation. Please specify the "
+                "task name, which should be corresponding to a registered evaluation "
+                "function, using the decorator `@register_evaluation_function`."
+            )
+
+        if evaluation_fn := REGISTRY.get_evaluation_function(task_params.task_name):
+            task_params.evaluation_fn = evaluation_fn
```
[the remark this answers is not in the document]
Comment (kaisopos) [reply src/oumi/core/evaluation/evaluator.py]: Dropped `CustomTaskParams`, as it's not needed (per your other comment)

File: src/oumi/core/configs/params/evaluation_params.py [hunk 14]
```diff
@@ -208,12 +212,8 @@ def __post_init__(self):
...
+    """Parameters for running custom Oumi evaluations."""
 
-    data: DatasetSplitParams = field(default_factory=DatasetSplitParams)
-    """Parameters for the dataset split to be used in evaluation.
-
-    This includes specifications for train, validation, and test splits,
-    as well as any data preprocessing parameters.
-    """
+    evaluation_fn: Optional[Callable] = None
+    """User-defined function to evaluate a model on the custom task."""
```
[the remark this answers is not in the document]
Comment (kaisopos) [reply src/oumi/core/configs/params/evaluation_params.py]: You are right, it's not useful, dropped it

File: src/oumi/core/configs/params/evaluation_params.py [hunk 15]
```diff
@@ -56,12 +56,12 @@ class EvaluationTaskParams(BaseParams):
...
             )
     """
 
-    evaluation_platform: str = MISSING
-    """The evaluation platform to use for the current task."""
+    evaluation_backend: str = MISSING
+    """The evaluation backend to use for the current task."""
 
     task_name: Optional[str] = None
     """The task to evaluate."""
```
[the remark this answers is not in the document]
Comment (kaisopos) [reply src/oumi/core/configs/params/evaluation_params.py]: Yes, missed that -- thank you!

File: src/oumi/core/configs/params/evaluation_params.py (hunk 7 above)
Comment (kaisopos) [reply src/oumi/core/configs/params/evaluation_params.py:38]: Added + one example

File: src/oumi/core/evaluation/backends/alpaca_eval.py (hunk 8 above)
Comment (kaisopos) [reply src/oumi/core/evaluation/backends/alpaca_eval.py]: Yes, great suggestion - thanks!

Comment (kaisopos) [reply src/oumi/core/evaluation/backends/alpaca_eval.py]: Fixed

File: src/oumi/core/evaluation/backends/lm_harness.py (hunk 9 above)
Comment (kaisopos) [reply src/oumi/core/evaluation/backends/lm_harness.py]: Fixed too!

**Review by taenin — APPROVED** [review]
Approving so you're not blocked here. Please address the above feedback before submission.

**Review by wizeng23 — APPROVED** [review]
LGTM! Please address any remaining comments before submission

File: src/oumi/core/configs/params/evaluation_params.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (kaisopos) [reply src/oumi/core/configs/params/evaluation_params.py]: SG - Moved it to the Evaluator as a private method.
Added more tests too.

File: src/oumi/core/configs/params/evaluation_params.py
Comment (kaisopos) [reply src/oumi/core/configs/params/evaluation_params.py:1]: Yes, I have searched exhaustively for this :)

File: src/oumi/core/evaluation/evaluation_result.py (hunk 10 above)
Comment (kaisopos) [reply src/oumi/core/evaluation/evaluation_result.py:18]: Let me skip this for now - I want to wrap this PR today

File: src/oumi/evaluation/save_utils.py [hunk 16]
```diff
@@ -41,57 +37,105 @@ def _save_to_file(output_path: Path, data: Any) -> None:
...
+    if not backend_name:
+        raise ValueError("The evaluation backend name must be provided.")
+    base_output_dir = base_output_dir or "."
+
     # Create the output directory: `<base_output_dir> / <platform>_<time>`.
-    output_dir = Path(base_output_dir) / f"{platform.value}_{start_time_str}"
-    output_dir.mkdir(parents=True, exist_ok=True)
+    start_time_in_path = f"_{start_time_str}" if start_time_str else ""
+    output_dir = Path(base_output_dir) / f"{backend_name}{start_time_in_path}"
+    if output_dir.exists():
```
[the remark this answers is not in the document]
Comment (kaisopos) [reply src/oumi/evaluation/save_utils.py:100]: Fixed

File: src/oumi/core/evaluation/evaluation_result.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (kaisopos) [reply src/oumi/core/evaluation/evaluation_result.py:18]: Done

File: src/oumi/core/evaluation/evaluator.py [hunk 17]
```diff
@@ -0,0 +1,185 @@
...
+from oumi.core.evaluation.backends.alpaca_eval import evaluate as evaluate_alpaca_eval
+from oumi.core.evaluation.backends.lm_harness import evaluate as evaluate_lm_harness
+from oumi.core.evaluation.evaluation_result import EvaluationResult
+from oumi.core.registry import REGISTRY
+from oumi.evaluation.platform_prerequisites import check_prerequisites
+from oumi.evaluation.save_utils import save_evaluation_output
+
+
+class Evaluator:
+    """Oumi evaluator for evaluating models on various backends and tasks."""
```
[the remark this answers is not in the document]
Comment (kaisopos) [reply src/oumi/core/evaluation/evaluator.py]: Done
